dali/util/s3_filesystem.cc

```diff
@@ -0,0 +1,125 @@
+
+namespace dali {
+
+namespace s3_filesystem {
+
+static const char kAllocationTag[] = "s3_filesystem";
+
+S3ObjectLocation parse_uri(const std::string& uri) {
+  auto parsed_uri = ParseURI(uri);
+  if (parsed_uri.scheme != "s3")
+    throw std::runtime_error("Not an S3 URI: " + uri);
+  S3ObjectLocation object_location;
+  object_location.bucket = parsed_uri.authority;
+  object_location.object = parsed_uri.path;
+  return object_location;
+}
+
+S3ObjectStats get_stats(Aws::S3::S3Client* s3_client, const S3ObjectLocation& object_location) {
+  DomainTimeRange tr(make_string("get_stats @ ", object_location.object), DomainTimeRange::kOrange);
+  S3ObjectStats stats;
+  if (object_location.object.empty())
+    throw std::runtime_error("Object can't be empty");
+
+  Aws::S3::Model::HeadObjectRequest head_object_req;
+  head_object_req.SetBucket(object_location.bucket.c_str());
+  head_object_req.SetKey(object_location.object.c_str());
+  head_object_req.SetResponseStreamFactory(
+      []() { return Aws::New<Aws::StringStream>(kAllocationTag); });
+  auto head_object_outcome = s3_client->HeadObject(head_object_req);
+  stats.exists = head_object_outcome.IsSuccess();
+  stats.size = stats.exists ? head_object_outcome.GetResult().GetContentLength() : 0;
+  return stats;
+}
+
+
+size_t read_object_contents(Aws::S3::S3Client* s3_client, const S3ObjectLocation& object_location,
+                            void* buf, size_t n, size_t offset) {
+  std::stringstream ss;
+  ss << "bytes=" << offset << "-" << offset + n - 1;
+  std::string byte_range_str = ss.str();
+
+  DomainTimeRange tr(
+      make_string("read_object_contents @ ", object_location.object, " ", byte_range_str),
+      DomainTimeRange::kOrange);
+
+  Aws::S3::Model::GetObjectRequest getObjectRequest;
+  getObjectRequest.SetBucket(object_location.bucket.c_str());
+  getObjectRequest.SetKey(object_location.object.c_str());
+  getObjectRequest.SetRange(byte_range_str.c_str());
+
+  Aws::Utils::Stream::PreallocatedStreamBuf streambuf(reinterpret_cast<uint8_t*>(buf), n);
+  getObjectRequest.SetResponseStreamFactory(
+      [&streambuf]() { return Aws::New<Aws::IOStream>(kAllocationTag, &streambuf); });
+
+  size_t bytes_read = 0;
+  auto getObjectOutcome = s3_client->GetObject(getObjectRequest);
+  if (!getObjectOutcome.IsSuccess()) {
+    const Aws::S3::S3Error& err = getObjectOutcome.GetError();
+    throw std::runtime_error(err.GetExceptionName() + ": " + err.GetMessage());
+  } else {
+    bytes_read = getObjectOutcome.GetResult().GetContentLength();
+  }
+  return bytes_read;
+}
+
+void list_objects_f(Aws::S3::S3Client* s3_client, const S3ObjectLocation& object_location,
+                    PerObjectCallable per_object_call) {
+  DomainTimeRange tr(make_string("list_object_contents @ ", object_location.object),
+                     DomainTimeRange::kOrange);
+  std::string prefix = object_location.object;
+  if (prefix.back() != '/') {
+    prefix.push_back('/');
+  }
```

Comment:
Could we abuse the prefix to also include a common prefix of the filters? E.g.:
```location = "./data"
filters = ["img*.jpg", "img*.png"]
```
would produce
```
prefix = "./data/img"
```

Reply:
no, I don't think so. The prefix seems to be like a "directory" path